apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/Numeric.cs

```diff
@@ -0,0 +1,663 @@
+
+        private static readonly ConcurrentDictionary<Parameters, NumericType> s_defined = new ConcurrentDictionary<Parameters, NumericType>();
+
+        private readonly Parameters _parameters;
+        private readonly Lazy<NumericValue> _lazyMaxValue;
+        private readonly Lazy<NumericValue> _lazyMinValue;
+        private readonly Lazy<NumericValue> _lazyEpsilon;
+        private readonly Lazy<NumericValue> _lazyZero;
+
+        private NumericType(Parameters parameters)
+        {
+            _parameters = parameters;
+            _lazyMinValue = new Lazy<NumericValue>(() => ValueFromScaledInteger(_parameters.ScaledMinValue));
+            _lazyMaxValue = new Lazy<NumericValue>(() => ValueFromScaledInteger(_parameters.ScaledMaxValue));
+            _lazyEpsilon = new Lazy<NumericValue>(() => ValueFromScaledInteger(1));
+            _lazyZero = new Lazy<NumericValue>(() => ValueFromScaledInteger(0));
+        }
+
+        internal static NumericType NumericTypeFor(BigInteger scaledMinValue, BigInteger scaledMaxValue, int scale)
+        {
+            Parameters parameters = new Parameters(scaledMinValue, scaledMaxValue, scale);
+            return s_defined.GetOrAdd(parameters, new NumericType(parameters));
+        }
+
+        internal static NumericType NumericTypeFor(BigInteger scaledMaxValue, int scale)
+        {
+            // Min is the negation of max
+            Parameters parameters = new Parameters(-scaledMaxValue, scaledMaxValue, scale);
+            return s_defined.GetOrAdd(parameters, new NumericType(parameters));
+        }
+
+        /// <summary>
+        /// Scale of this type, i.e. max number of digits after the decimal point.
+        /// </summary>
+        public int Scale { get => _parameters.Scale; }
+
+        /// <summary>
+        /// Zero represented as a <see cref="NumericValue"/>. This is the default value for <see cref="NumericValue"/>.
+        /// </summary>
+        public NumericValue Zero { get => _lazyZero.Value; }
+
+        /// <summary>
+        /// The maximum valid value for a <see cref="NumericValue"/> of this <see cref="NumericType"/>.
+        /// </summary>
+        public NumericValue MaxValue { get => _lazyMaxValue.Value; }
+
+        /// <summary>
+        /// The minimum valid value for <see cref="NumericValue"/> of this <see cref="NumericType"/>.
+        /// </summary>
+        public NumericValue MinValue { get => _lazyMinValue.Value; }
+
+        /// <summary>
+        /// The smallest positive value for <see cref="NumericValue"/> of this <see cref="NumericType"/>.
+        /// </summary>
+        public NumericValue Epsilon { get => _lazyEpsilon.Value; }
+
+        public override bool Equals(object obj) =>
+            obj is NumericType type && _parameters.Equals(type._parameters);
+
+        public override int GetHashCode() => _parameters.GetHashCode();
+
+        internal static NumericType GetUnionType(NumericType first, NumericType second)
```

Comment:
I'm not sure this is necessary - I suspect that just prohibiting the "mix and match" addition and subtraction would simplify things. (We do need to handle adding/subtracting the default value of course, but that's really simple.)